sync-android/src/main/java/com/cloudant/sync/datastore/encryption/KeyStorage.java

```diff
@@ -0,0 +1,186 @@
+                        (savedObject.getString(KEY_SALT)), DKPEncryptionUtil
+                        .hexStringToByteArray(savedObject.getString(KEY_IV)), savedObject.getInt
+                        (KEY_ITERATIONS), savedObject.getString(KEY_VERSION));
+            } catch (JSONException e) {
+                LOGGER.log(Level.SEVERE, e.getLocalizedMessage(), e);
+                return null;
+            }
+        }
+        return encryptionData;
+    }
+
+    /**
+     * Save to the {@link SharedPreferences} a {@link KeyData}.
+     * <p/>
+     * Notice that if there is already data in the {@link SharedPreferences} bound to the same
+     * identifier used to create
+     * this storage, the operation will fail.
+     *
+     * @param data {@link KeyData} to save to the {@link SharedPreferences}
+     * @return true (success) or false (fail)
+     */
+    public boolean saveEncryptionKeyData(KeyData data) {
+        JSONObject objectToSave = new JSONObject();
+        try {
+            objectToSave.put(KEY_DPK, DKPEncryptionUtil.byteArrayToHexString(data
+                    .getEncryptedDPK()));
+            objectToSave.put(KEY_IV, DKPEncryptionUtil.byteArrayToHexString(data.getIv()));
+            objectToSave.put(KEY_SALT, DKPEncryptionUtil.byteArrayToHexString(data.getSalt()));
+            objectToSave.put(KEY_ITERATIONS, data.getIterations());
+            objectToSave.put(KEY_VERSION, data.getVersion());
+
+            String valueToSave = objectToSave.toString();
+
+            SharedPreferences.Editor editor = this.prefs.edit();
+            editor.putString(buildKey(), valueToSave);
+            editor.commit();
+
+        } catch (JSONException e) {
+            LOGGER.log(Level.SEVERE, e.getLocalizedMessage(), e);
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Remove from the {@link SharedPreferences} a {@link KeyData} associated to the same
+     * identifier used to
+     * create this storage.
+     * <p/>
+     * It will succeed if the data is deleted or if there is no data at all.
+     *
+     * @return true (success) or false (fail)
+     */
+    public boolean clearEncryptionKeyData() {
+        try {
+            SharedPreferences.Editor editor = this.prefs.edit();
+            editor.remove(buildKey());
+            editor.commit();
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, e.getLocalizedMessage(), e);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Look for data saved in the {@link SharedPreferences} with the same identifier used to
+     * create this storage instance.
+     *
+     * @return true (data found) or false (data not found)
+     */
+    public boolean encryptionKeyDataExists() {
+        return this.prefs.contains(buildKey());
+    }
+
+    private String buildKey() {
+        return PREF_NAME_DPK + "-" + this.account;
+    }
```

Comment:
Changes requested in comment included in https://github.com/cloudant/sync-android/commit/dd276d263bfc9c5ddebb047cecd27230cb074f63